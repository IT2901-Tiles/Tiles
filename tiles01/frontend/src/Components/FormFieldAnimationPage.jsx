import React, { useState } from 'react'


function FormFieldAnimationPage(props) {
    const [input, setInput] = useState('')
    const [triggerCard] = useState(props.triggerCard) // props to pass to AnimationPage.jsx
    const [thingCard] = useState(props.thingCard) // props to pass to AnimationPage.jsx
    const [feedbackCard] = useState(props.feedbackCard) // props to pass to AnimationPage.jsx
    let outputStandardText = `Cards chosen:\n- Trigger: ${triggerCard}\n- Thing: ${thingCard}\n- Feedback: ${feedbackCard}\n\nYour idea:\n- ` // passed as first parameter in the downloadFile Blob

    // The following function gathers the content from the textarea and inserts it into a external txt-file
    function downloadFile() {
        let userInputValue // will change depending on the user writes something or not

        if (document.getElementById("textBoxAnimationPage").value === "") { // if user does not write anything in the textarea
            userInputValue = "Oops! Looks like no brilliant ideas were written down..."
        } else { // if the user writes something in the textare
            userInputValue = document.getElementById("textBoxAnimationPage").value
        }

        const htmlElement = document.createElement("a") // creates an "a" element in the html page, where we'll "store" the value of the user input
        const valueInput = new Blob([outputStandardText + userInputValue], { type: 'text/plain' }) // value of txt file must be passed in the first parameter of Blob
        htmlElement.href = URL.createObjectURL(valueInput)
        htmlElement.download = "tiles.txt" // name of the file that is being downloaded
        document.body.appendChild(htmlElement)
        htmlElement.click()

        // JSON-request
        sendIdeaRequest({"trigger":props.triggerCard,"thing":props.thingCard,"feedback":props.feedbackCard,"idea":userInputValue})
    }

    return (
        <div className="textAreaAnimationPage">
            {/* The text-area */}
            <textarea id="textBoxAnimationPage" placeholder="Write your idea here..." onChange={newInput => setInput(newInput.target.value)} />
            {/* Download button */}
            <button onClick={downloadFile} className="textButtonAnimationPage">Download</button>
        </div>
    )
}

// sendIdeaRequest fetches the idea (data) as a JSON request, is used in the downloadFile-function within the FormFieldAnimationPage function 
function sendIdeaRequest(data){
    fetch('/add_idea', {
      method: 'post',
      headers: {'Content-Type': 'application/json'},
      body: JSON.stringify(data)
    }).then(response=>response.json()).then(data=>{ 
        window.alert(data) 
    })
}

export default FormFieldAnimationPage