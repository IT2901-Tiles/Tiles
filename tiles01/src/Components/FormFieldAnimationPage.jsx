import React, { useState } from 'react'


function FormFieldAnimationPage(props) {
    const [input, setInput] = useState('')
    const [triggerCard] = useState(props.triggerCard) // props to pass to AnimationsPreview.jsx
    const [thingCard] = useState(props.thingCard) // props to pass to AnimationsPreview.jsx
    const [feedbackCard] = useState(props.feedbackCard) // props to pass to AnimationsPreview.jsx
    let outputStandardText = `Cards chosen:\n- Trigger: ${triggerCard}\n- Thing: ${thingCard}\n- Feedback: ${feedbackCard}\n\nYour idea:\n- ` // passed as first parameter in the downloadFile Blob

    // The following function gathers the content from the textarea and inserts it into a external txt-file
    function downloadFile() {
        let userInputValue // will change depending on the user writes something or not 
        if (document.getElementById("textBoxAnimationPage").value === "") { // if user does not write anything in the textarea
            userInputValue = "Oops! Looks like no brilliant ideas were written down..." 
        } else { // if the user writes something in the textare
            userInputValue = document.getElementById("textBoxAnimationPage").value
        }
        const htmlElement = document.createElement("a") // creates an "a" element in the html page, where we'll "store" the value of the user input
        const valueInput = new Blob([outputStandardText + userInputValue], {type: 'text/plain'}) // value of txt file must be passed in the first parameter of Blob
        htmlElement.href = URL.createObjectURL(valueInput)
        htmlElement.download = "tiles.txt"
        document.body.appendChild(htmlElement)
        htmlElement.click()
    }

    // Will fix the position when the animations are ready
    return (
        <div className="textAreaAnimationPage">
            <textarea id="textBoxAnimationPage" placeholder="Write your idea here..." onChange={newInput => setInput(newInput.target.value)}/>
                <button onClick={downloadFile} className="textButtonAnimationPage">Save</button>
        </div>
    );
}

export default FormFieldAnimationPage