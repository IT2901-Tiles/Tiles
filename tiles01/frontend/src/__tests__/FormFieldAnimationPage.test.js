import {configure, shallow} from "enzyme"
import Adapter from 'enzyme-adapter-react-16';
import FormFieldAnimationPage from '../Components/FormFieldAnimationPage';
import {screen, cleanup, render} from '@testing-library/react';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';

afterEach(() => {
    cleanup();
});

configure({
    adapter: new Adapter()
});

describe('FormFieldAnimationPage component', () => {
    const wrapper = shallow(<FormFieldAnimationPage/>);
    it('renders correctly', () => {
        const tree = renderer.create(<FormFieldAnimationPage/>);
        expect(tree).toMatchSnapshot();
    });
    it('renders a "Save"-button element', () => {
        expect(wrapper.find('button')).toHaveLength(1);
        expect(wrapper.find('button').text()).toEqual('Save');
    });
    it('renders a textarea', () => {
        expect(wrapper.find('textarea')).toHaveLength(1);
    });
    it('button calls a function when clicked', () => {
        const mockedDownloadFile = jest.fn();
        render(<button onClick={mockedDownloadFile} className="textButtonAnimationPage">Save</button>);
        userEvent.click(screen.getByRole('button'));
        expect(mockedDownloadFile).toHaveBeenCalledTimes(1);
    });
    it('textarea calls a function on change', () => {
        const mockedSetInput = jest.fn();
        render(<textarea id="textBoxAnimationPage" placeholder="Write your idea here..." onChange={mockedSetInput} />);
        userEvent.type(screen.getByRole('textbox'), 'My great idea');
        expect(mockedSetInput).toHaveBeenCalledTimes(13);
    });  
});
