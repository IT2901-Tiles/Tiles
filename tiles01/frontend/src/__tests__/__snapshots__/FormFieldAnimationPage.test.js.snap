// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFieldAnimationPage component renders correctly 1`] = `
<div
  className="textAreaAnimationPage"
>
  <textarea
    id="textBoxAnimationPage"
    onChange={[Function]}
    placeholder="Write your idea here..."
  />
  <button
    className="textButtonAnimationPage"
    onClick={[Function]}
  >
    Save
  </button>
</div>
`;
